### ISSUE TITLE
Revamp listener support in Salesforce connector

### ISSUE THREAD

**Issue by aashikam** [ballerina-platform/ballerina-library#6177]
**Description:**
The salesforce connector currently supports the listener implementation through the `ballerinax/trigger.salesforce` package. Ideally this should be supported through the `ballerinax/salesforce` package itself.

**Comment by aashikam** [comment]
# Update

1. The Salesforce listener is moved to the Salesforce package with minor updates to the exiting code bases. This is supported under the salesforce.events module. There are few couple of build script issues to be fixed. 

2. There is a new API available called [Pub/Sub API](https://developer.salesforce.com/docs/platform/pub-sub-api/guide/intro.html), which provides a single interface for publishing and subscribing to platform events.

## Streaming API (Existing one)
Allows you to receive updates or events from Salesforce in real-time without needing to constantly ask for them. Instead of your application repeatedly asking Salesforce if there are any updates, Salesforce sends the updates directly to your application as they happen.

- How it works: Salesforce "pushes" the updates to your application as they occur. 

- Why it's useful:
    It saves resources because your application doesn't have to keep asking for updates.
    It's faster because updates are sent immediately when they occur.
    It's more efficient because it reduces unnecessary requests to Salesforce.

- How clients connect:
    Connects to Salesforce using the HTTP/1.1 request-response model and the Bayeux protocol, which is implemented using CometD.

- Message reliability and durability:
    Salesforce ensures that messages are reliably delivered to clients, even if they were disconnected temporarily.
    It stores different types of events for a certain period, allowing clients to retrieve missed events when they reconnect.

- Types of events:
Salesforce offers different types of events, such as PushTopic events, generic events, platform events, and Change Data Capture events. Each type of event has its own features and use cases.
In simple terms, the Streaming API lets your application receive updates from Salesforce in real-time, which is faster, more efficient, and saves resources compared to constantly asking for updates. It's like getting instant notifications whenever something important happens, without having to keep checking for them manually.

## Pub/Sub API
 Lets you easily send and receive messages about events happening in a system. These events could be anything from real-time updates to changes in data.

- What it does: The Pub/Sub API allows you to both send (publish) and receive (subscribe) these events using a single interface.

- Who can use it: It's available for certain editions of Salesforce like Enterprise, Performance, Unlimited, and Developer Editions, but not for Government Cloud.

- Why it's useful:

    It simplifies your work by handling both publishing and subscribing in one place.
    It ensures that messages are successfully sent or received, without any loss.
    It lets you control how many events you receive at once, depending on how fast your system can handle them.
    It's secure and can handle a large volume of events efficiently.

- How it works:

    It uses gRPC API along with HTTP/2 for fast and efficient communication.
    It supports multiple programming languages like Python, Java, Node.js, etc.
    It allows bidirectional communication, meaning both the sender and receiver can talk to each other.
    It integrates seamlessly across different Salesforce services (like Sales Cloud, Service Cloud, Marketing Cloud, etc.) and even with external systems.


### Similarities:

1. Real-time event handling: Both APIs support real-time event handling, allowing applications to receive updates as they occur without needing to constantly poll for them.
2. Subscription mechanism: Both APIs provide a subscription mechanism where clients can subscribe to specific types of events they're interested in.
3. Reduction in unnecessary requests: Both APIs aim to reduce unnecessary requests to the server by delivering updates only when there is new information available.
3. Efficiency: Both APIs are designed to be efficient in delivering updates to clients, improving the overall performance of applications that rely on real-time data.

### Differences:

1. Technology stack: The Pub/Sub API is based on gRPC API and HTTP/2, while the Streaming API uses HTTP/1.1 and the Bayeux protocol (CometD implementation). This difference in underlying technologies might impact the implementation details and integration with different systems.
2. Event types: The Pub/Sub API primarily focuses on publishing and subscribing to platform events and change data capture events, whereas the Streaming API supports a broader range of event types including PushTopic events, generic events, platform events, and Change Data Capture events.
3. Message reliability: The Pub/Sub API provides reliable message delivery by enabling replay of past events through durable streaming. The Streaming API offers similar functionality for clients subscribed with API version 37.0 or later. 
4. Message durability: Salesforce stores different types of events for different durations, and the mechanisms for retrieving missed events differ between the two APIs. The Pub/Sub API allows retrieval of events within the retention window through durable streaming, whereas the Streaming API stores events for a certain period and allows clients to retrieve missed events when they reconnect.

Considering the implementation and integration complexity, this migration would take some time as this is entirely new. Can't guarantee there wont be any breaking changes either.

**Comment by aashikam** [comment]
## Streaming API vs Pub/Sub API

### Streaming API

The Salesforce Streaming API utilizes the Bayeux Protocol and the CometD framework to enable real-time event streaming and communication between Salesforce servers and client applications.

The Salesforce Streaming API primarily operates using HTTP/1.1, with support for HTTP/2 in some cases, depending on the client and server configurations.

- Works on the CometD protocol and uses long-lived connections to push events from Salesforce to client applications.
- Suitable for scenarios where you need real-time updates and want to receive a continuous stream of events.
- Good for applications that require low-latency updates and high-volume data processing.
- Well-suited for use cases such as monitoring, real-time analytics, and live dashboards.
- Filtering in the Streaming API is typically done through the use of SOQL (Salesforce Object Query Language) queries called "PushTopic queries."
- When you subscribe to a PushTopic, you can specify criteria in the form of a SOQL query that defines which events you want to receive.
- The Streaming API then only sends events that match the criteria defined in the SOQL query.
- If you are already familiar with Salesforce and its query language SOQL, using the Streaming API might feel more straightforward.
- It has a more direct integration with Salesforce objects and events, making it easier to understand for developers who are already comfortable with Salesforce data structures and operations.

### Pub/Sub API

A newer API offering similar capabilities to the Streaming API but based on gRPC API and HTTP/2, providing efficient binary event messages., delivering binary event messages in the Apache Avro format, ensuring efficient publishing and delivery.
- Uses a publish-subscribe model where clients subscribe to specific channels/topics to receive events.
- Provides more flexibility in terms of event filtering and subscribing to specific types of events.
- Suitable for scenarios where you need more control over the types of events you receive and want to reduce the volume of data transmitted.
- Can be more efficient for applications that don't require constant real-time updates but still need to react to specific events.
- Filtering in the Pub/Sub API is typically done at the subscription level.
- When you subscribe to a topic/channel, you can specify filtering criteria such as event types, attributes, or other conditions.
- The Pub/Sub API then only delivers events to your subscription that match the specified criteria.
- The Pub/Sub model, based on topics and subscriptions, is a common pattern in messaging systems and may be more familiar to developers who have experience with messaging architectures.
- It offers more flexibility in terms of event filtering and subscription management, which can be advantageous for complex use cases.

### Considerations
- In terms of raw simplicity, the Streaming API might be slightly easier to get started with, especially for developers already accustomed to working with Salesforce. 
- Conceptual Simplicity:
The Pub/Sub model, based on topics and subscriptions, aligns closely with common messaging patterns used in software development.
Topics represent the channels through which events are published, and subscriptions define the interest in specific topics.
This model can be easier to grasp conceptually, especially if you have experience with messaging systems or event-driven architectures.
- Flexibility in Filtering:
The Pub/Sub API offers flexibility in filtering events at the subscription level, allowing you to specify criteria such as event types or attributes without needing to write SQL-like queries.
This approach can be more intuitive for developers who are accustomed to working with filters or conditions in other programming contexts.

In summary, if the application requires real-time updates and continuous streaming of events, the Streaming API might be a better choice. However, if you need more flexibility in event filtering and control over the types of events received, the Pub/Sub API could be a better fit.

**Comment by aashikam** [comment]
# Design Review: Ballerina Salesforce Listener

## Overview

The `ballerinax/trigger.salesforce` module provides a Listener to grasp events triggered from a Salesforce org. This functionality is provided by [Salesforce Streaming API](https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/intro_stream.htm).

## Prerequisites
Before using this connector in your Ballerina application, complete the following:

- Create a Salesforce account.
- Obtain the security token for your Salesforce org by visiting <br/> `Profile -> Settings -> My Personal Information -> Reset My Security Token`
- Add the **API Enabled** permission & **Streaming API** permissions for your Salesforce org. 
- Prerequisites for different kinds of events
    - Subscribe to **Change Data Capture Events**
        1. Subscribe to channels to receive notifications for record changes by visiting <br/> `Setup -> Integrations -> Change Data Capture`

## Quickstart
To use the Salesforce listener in your Ballerina application, update the .bal file as follows:

### Step 1: Import listener
Import the `ballerinax/trigger.salesforce as sfdc` module as shown below.
```ballerina
import ballerinax/trigger.salesforce as sfdc;
```

### Step 2: Create a new listener instance
Create a `sfdc:ListenerConfig` configuration object with your `Salesforce User Name`, `Salesforce Password`, `Salesforce Security Token`, and `Subscribe Channel Name`. Then, initialize the listener with it. 

Notes: 
- For using the listener in sandbox environments, specify the parameter `environment: "Sandbox"`.

```ballerina
sfdc:ListenerConfig configuration = {
    username: "USER_NAME",
    password: "PASSWORD" + "SECURITY_TOKEN",
    channelName: "CHANNEL_NAME"
};
listener Listener sfdc:Listener = new (configuration);
```

### Step 3: Implement a listener remote function
Now, implement a listener service with remote functions to handle specific event types.

```ballerina
import ballerina/log;
import ballerinax/trigger.salesforce as sfdc;

service sfdc:RecordService on sfdcListener {
    remote function onUpdate(sfdc:EventData event) returns error? {
        log:printInfo(event.toString());
    }
    remote function onCreate(sfdc:EventData event) returns error? {
    }
        
    remote function onDelete(sfdc:EventData event) returns error? {
    }
    remote function onRestore(sfdc:EventData event) returns error? {
    }
}
```

### Step 4: Compile and run
Use `bal run` command to compile and run the Ballerina program.

* Receiving events
    * After successful verification of Request URL your Ballerina service will receive events.

## Discussion Points

- Should the initialization and configuration of the listener be handled similarly to client initialization and configuration?

This guide provides a quick overview of how to set up and use the Ballerina Salesforce Listener. For detailed information and additional functionalities, please refer to the official [Salesforce Streaming API documentation](https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/intro_stream.htm) and the [Ballerina Salesforce Listener documentation](https://ballerina.io/learn/api-docs/ballerinax/trigger.salesforce.html).

**Comment by aashikam** [comment]
## The changes proposed at the design review

1. Move the listener in to the salesforce default package, so the listener would be available as `salesforce:Listener`.
2. Changes to Listener configuration and init:
```ballerina

# Configurations related to authentication.
#
# + username - Username to use for authentication
# + password - Password/secret/token to use for authentication
public type Credentials record {|
    string username;
    string password;
|};

# Salesforce listener configuration.
# 
# + auth - Configurations related to username/password authentication
# + replayFrom - The replay ID to change the point in time when events are read
# + isSandbox - The type of salesforce environment, if sandbox environment or not
public type ListenerConfig record {|
    Credentials auth;
    int|ReplayOpitons replayFrom?;
    boolean isSandbox = false;
|};

public enum ReplayOpitons {
   REPLAY_FROM_TIP,
   REPLAY_FROM_EARLIEST
}
```
3. Move subscription to service level, currently there should be one topic subscription per listener but this can be changed to one subscription per service. Therefore the listener can be shared among the services. The topic/channel name can be given as the service name. 
```ballerina
service "data/ChangeEvents" on eventListener {
    remote function onCreate(EventData payload) {
        io:println("Created " + payload.toString());
    }
}
```

5. Changes to remote functions 
These are the generic structures of the event types. 
   - create
```json
{
  "eventType": "created",
  "objectType": "Account",
  "recordId": "001XXXXXXXXXXXX",
  "fields": {
    "Name": {
      "oldValue": null,
      "newValue": "New Account Name"
    },
    "Industry": {
      "oldValue": null,
      "newValue": "Technology"
    },
    // Other fields that were modified
  }
}
```
  - update
```json
{
  "eventType": "updated",
  "objectType": "Contact",
  "recordId": "003XXXXXXXXXXXX",
  "fields": {
    "FirstName": {
      "oldValue": "John",
      "newValue": "Jane"
    },
    "LastName": {
      "oldValue": "Doe",
      "newValue": "Smith"
    },
    // Other fields that were modified
  }
}

```
  - delete
```json
{
  "eventType": "deleted",
  "objectType": "Opportunity",
  "recordId": "006XXXXXXXXXXXX",
  "fields": {
    // No fields will be present in a delete event
  }
}

```
  - restore
```json
{
  "eventType": "undeleted",
  "objectType": "Case",
  "recordId": "500XXXXXXXXXXXX",
  "fields": {
    // No fields will be present in an undelete event
  }
}
```
Can add these types as the first parameter and the ChangeEventMetadata as the second optional parameter.

**Comment by niveathika** [comment]
+1 to the above changes. I have added a few more suggestions

1. Listener inialisation using included param of the `ListenerConfig`
```ballerina
listener Listener sfdc:Listener = new (*sfdc:ListenerConfig);
```

Lets update listener initalisation as above, This ensures user only have to give auth details

```
listener Listener sfdc:Listener = new (auth = {username = "", password = ""});
```
2. Rename `Credentials` record to `CredentialsConfig` to be consistent with other libs.

**Comment by aashikam** [comment]
This was de-prioritized due to support tasks and other feature requests. We have planned this for the upcoming sprint.
